src/backend/distributed/commands/multi_copy.c

```diff
@@ -986,6 +1023,49 @@ OpenCopyTransactions(CopyStmt *copyStatement, ShardConnections *shardConnections
 
 
 /*
+ * CanUseBinaryCopyFormat iterates over columns of the relation given in rowOutputState
+ * and looks for a column whose type is array of user-defined type. If it finds that
+ * means we cannot use binary format for COPY and it returns false.
+ */
+static bool
+CanUseBinaryCopyFormat(TupleDesc tupleDescription, CopyOutState rowOutputState)
+{
+	bool useBinaryCopyFormat = true;
+	int totalColumnCount = tupleDescription->natts;
+	int columnIndex = 0;
+
+	for (columnIndex = 0; columnIndex < totalColumnCount; columnIndex++)
+	{
+		Form_pg_attribute currentColumn = tupleDescription->attrs[columnIndex];
+		Oid typeId = InvalidOid;
+		char typeCategory = '\0';
+		bool typePreferred = false;
+		bool builtInType = false;
+		bool containsOid = false;
+
+		if (currentColumn->attisdropped)
+		{
+			continue;
+		}
+
+		typeId = currentColumn->atttypid;
+		builtInType = typeId < FirstNormalObjectId;
+
+		get_type_category_preferred(typeId, &typeCategory, &typePreferred);
+		containsOid = typeCategory == TYPCATEGORY_ARRAY ||
+					  typeCategory == TYPCATEGORY_COMPOSITE;
```

Comment:
We shouldn't perform these tests when it's a builtin type. They're not all that cheap.